tracer/src/Datadog.Trace/DatabaseMonitoring/DatabaseMonitoringPropagator.cs

```diff
@@ -31,80 +31,109 @@ internal static class DatabaseMonitoringPropagator
-                 && integrationId is not (IntegrationId.SqlClient or IntegrationId.Oracle))
-                {
-                    traceParentInjected = true;
-                    propagatorStringBuilder.Append(',').Append(W3CTraceContextPropagator.TraceParentHeaderName).Append("='").Append(W3CTraceContextPropagator.CreateTraceParentHeader(span.Context)).Append("'*/");
-                }
-                else
-                {
-                    propagatorStringBuilder.Append("*/");
-                }
+            var traceParentInjected = false;
+            // For SqlServer & Oracle we don't inject the traceparent to avoid affecting performance, since those DBs generate a new plan for any query changes
+            if (propagationLevel == DbmPropagationLevel.Full
+             && integrationId is not (IntegrationId.SqlClient or IntegrationId.Oracle))
+            {
+                traceParentInjected = true;
+                propagatorStringBuilder.Append(',').Append(W3CTraceContextPropagator.TraceParentHeaderName).Append("='").Append(W3CTraceContextPropagator.CreateTraceParentHeader(span.Context)).Append('\'');
+            }
 
-                return StringBuilderCache.GetStringAndRelease(propagatorStringBuilder);
+            propagatorStringBuilder.Append("*/");
+
+            // modify the command to add the comment
+            var commandText = command.CommandText ?? string.Empty;
+            var propagationComment = StringBuilderCache.GetStringAndRelease(propagatorStringBuilder);
+            if (ShouldAppend(integrationId, commandText))
+            {
+                command.CommandText = $"{commandText} {propagationComment}";
+            }
+            else
+            {
+                // prepending the propagation comment is the preferred way,
+                // as this protects it from being truncated by the character limit if the command is very long.
+                command.CommandText = $"{propagationComment} {commandText}";
             }
 
-            return string.Empty;
+            return traceParentInjected;
+        }
+
+        internal static bool ShouldAppend(IntegrationId integrationId, string commandText)
+        {
+            // pg_hint_plan allows setting hints for the execution plan as the *first* comment. If such a hint is present,
+            // we need to append our comment rather than prepend, to avoid invalidating the hint
+            // see https://pg-hint-plan.readthedocs.io/en/latest/hint_details.html#syntax-and-placement
+            return integrationId == IntegrationId.Npgsql && StartsWithHint(commandText);
+        }
+
+        /// <summary>
+        /// Detect if a command contains a pg_hint_plan.
+        /// </summary>
+        private static bool StartsWithHint(string commandText)
+        {
+            // using .AsSpan() prevents from allocating a new string when we use TrimStart
+            return commandText.AsSpan().TrimStart().StartsWith(PgHintPrefix);
```

Comment:
Seems weird that `PgHintPrefix` had to be a `char[]` instead of a `string`, but I trust you tried it 😅  Maybe a side-effect of using a "vendored" `Span<T>`?

Reply:
Yeah idk why, but we have auto-cast from char[] but not from string 🤷